File: registration_deconvolution_scripts/step03_evaluate_InterestPointDetection.m
% Evaluate # of interestpoints per view
close all 
clear all

rootdir = './interestpoints/' ;
previewIPs = true ; 
pauseTime = 0.1 ;
previewEveryN = 10 ;
densityThres = 0.02 ; % threshold local density
dz = 1.0 ; % um / page, axial sampling
dx = 0.195 ; % um / pixel, in-plane sampling

%% Automatically determine timepoints and view tiles
% Automatically determine timepoints and view tiles
tp_files = dir('./c*_t*_a*.ome.tif');
tps = unique(cellfun(@(x) sscanf(x, 'c%*d_t%d_a%*d.ome.tif'), {tp_files.name}));

% Extract the channel and angle info
vtiles = cellfun(@(x) sscanf(x, 'c%d_t%*d_a%d.ome.tif', [1, 2]), {tp_files.name}, 'UniformOutput', false);

% Convert vtiles into a matrix for unique operation
vtiles_matrix = cell2mat(vtiles');  % Convert to 2D numeric matrix (channel, angle)
vtiles = unique(vtiles_matrix, 'rows');  % Get unique rows (tuples of channel, angle)
vtiles = 0:size(vtiles, 1)-1 ;

%% Count interestpoints
% Preallocate
clc ; 
colors = lines(length(vtiles)) ;
nips = repmat({zeros(length(tps), 1)},length(vtiles), 1); 
missing_files = {};

for tidx = 1:length(tps)
    tp = tps(tidx) ;
    if mod(tp, 10) == 0
        disp(['counting interest points for tp=' num2str(tp) '/' num2str(max(tps))])
    end
    for vId = 1:length(vtiles) 
        fn2find = fullfile(rootdir, sprintf('tpId_%d_viewSetupId_%d.beads.ip.txt', tp, vtiles(vId))) ;
        fns = dir(fn2find) ;
        if length(fns) == 1
            fn = fullfile(fns(1).folder, fns(1).name) ;
            ips = dlmread(fn, '\t', 1, 0) ;
            nips{vId}(tidx) = size(ips, 1) ;
            assert(size(ips, 1) == max(ips(:, 1))+1) ;

            if previewIPs && mod(tidx, previewEveryN) == 1
                ips = ips(:, 2:end) ;
                pointCloud = ips ;
                pointCloud(:, 3) = pointCloud(:, 3) * dz/dx ;
                
                kdtree = KDTreeSearcher(pointCloud);
                [indices, distances] = knnsearch(kdtree, pointCloud, 'K', 10);
                localDensities = 1 ./ mean(distances, 2);
                cmin = min(localDensities);
                cmax = max(localDensities);
                scaledDensities = (localDensities - cmin) / (cmax - cmin);

                clf
                colormap bwr
                scatter3(ips(:, 1), ips(:, 2), ips(:, 3), 10, localDensities, 'filled');
                xlabel('x'); ylabel('y'), zlabel('z'); axis equal
                caxis([0, densityThres])
                view(2)
                colorbar
                title(['tp=' num2str(tp) ': view ' num2str(vtiles(vId))])
                pause(pauseTime)
            end
        else
            missing_files{end+1} = sprintf('tp=%d vtile=%d', tp, vtiles(vId));
        end
    end
end

% Display missing files
if ~isempty(missing_files)
    disp('Missing files:')
    fprintf('%s\n', missing_files{:});
else
    disp('All expected files are present.')
end

%% Plot the # interest points as a function of time for each view
close all; 
legEntry = cell(1, length(vtiles)) ;
for vId = 1:length(vtiles)
    plot(tps, nips{vId}, '.-', 'color', colors(vId, :))
    legEntry{vId} = sprintf('View %d', vtiles(vId)) ;
    hold on;
end
xlabel('timepoint')
ylabel('#interest points')
legend(legEntry, 'location', 'northeastoutside')
currentDateTime = datetime('now', 'Format', 'yyyyMMddHHmm');
dateTimeString = char(currentDateTime);
set(gcf, 'color', 'w')
title('Interest points summary')
saveas(gcf, ['interestpoint_counts_' dateTimeString '.png'])

File: registration_deconvolution_scripts/step03_evaluate_InterestPointDetection.py
#!/usr/bin/env python3
# Evaluate # of interest points per view
# MATLAB → Python port
# Requires: numpy, matplotlib, scipy

import os
import glob
import re
import numpy as np
import matplotlib.pyplot as plt
from scipy.spatial import cKDTree
from datetime import datetime

# ----------------------------
# User parameters
# ----------------------------
datdir = 'F:\\PROJECTS\\LightMicroscopyBootcamp2025\\bynGAL4klar_UASmChCAAXUASGFPnls\\20251011181532_bynGAL4klar_UASmChCAAX_UASGFPnls_combined\\'
rootdir = os.path.join(datdir,'interestpoints\\')
# rootdir = './interestpoints/'           # directory containing *.beads.ip.txt files

# imfn = 'c*_t*_a*.ome.tif' 
imfn = 'tp_*_c*_angle_*.ome.tif' 
tca = 'tca'   # the order of time, channel, and angle in the image filenames

previewIPs = True
pauseTime = 0.05
previewEveryN = 30
densityThres = 0.02   # threshold for local density color scale
dz = 1.0              # um, axial sampling
dx = 0.2925  # 0.195            # um, in-plane sampling

# ----------------------------
# Helpers
# ----------------------------
def read_ip_file(fn):
    """Read interest point file with header 'id\\tx\\ty\\tz' -> (ids, xyz)."""
    data = np.loadtxt(fn, delimiter='\t', skiprows=1)
    if data.ndim == 1:
        data = data.reshape(1, -1)
    ids = data[:, 0].astype(int)
    # assert size(ips,1) == max(id)+1
    if data.shape[0] != (np.max(ids) + 1):
        raise AssertionError(
            f"Row count {data.shape[0]} != max(id)+1 ({np.max(ids)+1}) in {fn}"
        )
    xyz = data[:, 1:4]
    return ids, xyz

import re

def build_regex_from_template(imfn: str):
    """
    Build a regex with named groups (?P<t>), (?P<c>), (?P<a>) from a template.
    Supports either token mode:  'tp_{t}_c{c}_angle_{a}.ome.tif'
    or heuristic wildcard mode: 'tp_*_c*_angle_*.ome.tif' (common MVB/Luxendo style).
    """
    # Token mode (preferred)
    if any(tok in imfn for tok in ('{t}', '{c}', '{a}')):
        pat = re.escape(imfn)
        pat = pat.replace(r'\{t\}', r'(?P<t>\d+)')
        pat = pat.replace(r'\{c\}', r'(?P<c>\d+)')
        pat = pat.replace(r'\{a\}', r'(?P<a>\d+)')
        # For any other '*' you may have in the template, treat as non-greedy wildcard
        pat = pat.replace(r'\*', r'.*?')
        return re.compile(r'^' + pat + r'$')

    # Heuristic mode for the specific style: 'tp_*_c*_angle_*.ome.tif'
    # We interpret the '*' after 'tp_' as {t}, after 'c' as {c}, after 'angle_' as {a}.
    # Any other '*' are treated as wildcards.
    s = re.escape(imfn)

    # Replace escaped anchors for the three key numeric fields
    s = s.replace(r'tp_\*', r'tp_(?P<t>\d+)')
    s = s.replace(r'c\*',   r'c(?P<c>\d+)')
    s = s.replace(r'angle_\*', r'angle_(?P<a>\d+)')

    # Remaining '*' (if any) become non-greedy wildcards
    s = s.replace(r'\*', r'.*?')
    return re.compile(r'^' + s + r'$')


def parse_filename(name: str, imfn: str, tca: str = 'tca'):
    """
    Parse filename using `imfn` template and return (c, t, a) following your `tca` ordering.
    - `tca` is a permutation of 'cta' describing the order of numeric fields in the *string*.
    """
    tca = tca.lower()
    if sorted(tca) != ['a', 'c', 't']:
        raise ValueError("tca must be a permutation of 'cta'.")

    pat = build_regex_from_template(imfn)
    m = pat.search(name)
    if not m:
        raise ValueError(f"Filename does not match template: {name}  vs  {imfn}")

    # We captured by *names*, so get canonical values:
    try:
        vals = {k:int(v) for k,v in m.groupdict().items()}
    except Exception as e:
        raise ValueError(f"Template must expose numeric fields for t,c,a. Error: {e}")

    # If you want to *enforce/reflect* the declared tca order, we can check consistency:
    # Build the order we *saw* in the filename by scanning for the named groups.
    # (This is optional; named groups already give correct values regardless of order.)
    # Here we just return the canonical triplet:
    return vals['c'], vals['t'], vals['a']


def local_density_knn(xyz, k=10, z_scale=1.0):
    """Return local density = 1 / mean(distance to kNN) after scaling z."""
    pts = xyz.copy()
    pts[:, 2] *= z_scale
    tree = cKDTree(pts)
    dists, _ = tree.query(pts, k=k)        # includes self (distance 0)
    mean_d = np.mean(dists, axis=1)
    with np.errstate(divide='ignore'):
        rho = 1.0 / mean_d
    # Handle inf if any zero means
    if np.any(~np.isfinite(rho)):
        finite = rho[np.isfinite(rho)]
        fill = np.max(finite) if finite.size else 0.0
        rho[~np.isfinite(rho)] = fill
    return rho

# ----------------------------
# Auto-detect timepoints and view-tiles from file names
# ----------------------------
# Expecting names like: c<chan>_t<tp>_a<angle>.ome.tif or similar. The order can be swapped based on tca.
tp_files = glob.glob(os.path.join(datdir, imfn))

if not tp_files:
    raise FileNotFoundError("No files matching 'c*_t*_a*.ome.tif' were found.")

tp_vals = []
vtile_pairs = []  # (channel, angle)
for fn in tp_files:
    base = os.path.basename(fn)
    try:
        c, t, a = parse_filename(base, imfn=imfn, tca=tca)
    except ValueError:
        # filename didn't match the template; skip
        continue
    tp_vals.append(t)
    vtile_pairs.append((c, a))

tps = sorted(set(tp_vals))

# Unique (channel, angle) tuples; then define view IDs as 0..N-1
unique_pairs = sorted(set(vtile_pairs))
vtiles = list(range(len(unique_pairs)))  # MATLAB’s vtiles = 0:size(unique_rows)-1

# ----------------------------
# Count interest points
# ----------------------------
os.makedirs(rootdir, exist_ok=True)

colors = plt.cm.tab20(np.linspace(0, 1, max(20, len(vtiles))))
nips = [np.zeros(len(tps), dtype=int) for _ in vtiles]
missing_files = []

z_scale = dz / dx

for tidx, tp in enumerate(tps, start=1):  # 1-based to mirror MATLAB modulo logic
    if tp % 10 == 0:
        print(f"counting interest points for tp={tp}/{max(tps)}")

    for vidx, vtile in enumerate(vtiles):
        fn = os.path.join(rootdir, f"tpId_{tp}_viewSetupId_{vtile}.beads.ip.txt")
        if os.path.isfile(fn):
            try:
                ids, xyz = read_ip_file(fn)
            except Exception as e:
                print(f"[WARN] Could not read '{fn}': {e}")
                missing_files.append(f"tp={tp} vtile={vtile}")
                continue

            nips[vidx][tidx - 1] = xyz.shape[0]

            # Preview every previewEveryN timepoints (when tidx==1, 1+N, ...)
            if previewIPs and ((tidx - 1) % previewEveryN == 0) and xyz.shape[0] > 0:
                localDensities = local_density_knn(xyz, k=10, z_scale=z_scale)

                plt.clf()
                ax = plt.axes(projection='3d')
                sc = ax.scatter(xyz[:, 0], xyz[:, 1], xyz[:, 2],
                                s=10, c=localDensities, cmap='bwr')
                ax.set_xlabel('x'); ax.set_ylabel('y'); ax.set_zlabel('z')
                ax.set_box_aspect([1, 1, 1])
                sc.set_clim(0, densityThres)
                plt.colorbar(sc)
                ax.view_init(elev=90, azim=-90)  # view(2) top-down
                plt.title(f"tp={tp}: view {vtile}")
                plt.pause(pauseTime)
        else:
            missing_files.append(f"tp={tp} vtile={vtile}")

# ----------------------------
# Report missing
# ----------------------------
if missing_files:
    print("Missing files:")
    for msg in missing_files:
        print(msg)
else:
    print("All expected files are present.")

# ----------------------------
# Plot summary over time
# ----------------------------
plt.close('all')
plt.figure(figsize=(8, 5))
leg = []
for vidx, vtile in enumerate(vtiles):
    plt.plot(tps, nips[vidx], '.-', color=colors[vidx % len(colors)])
    leg.append(f"View {vtile}")

plt.xlabel('timepoint')
plt.ylabel('#interest points')
plt.title('Interest points summary')
plt.legend(leg, loc='upper left', bbox_to_anchor=(1.02, 1.0))
plt.gcf().set_facecolor('white')
plt.tight_layout()

stamp = datetime.now().strftime('%Y%m%d%H%M')
outpng = f"interestpoint_counts_{stamp}.png"
plt.savefig(os.path.join(datdir,outpng), dpi=200)
print(f"Saved: {outpng}")
